src/loaders/zarr/lib/storage.ts

```diff
@@ -0,0 +1,119 @@
+import { KeyError } from 'zarr';
+import type { AsyncStore } from 'zarr/dist/types/storage/types';
+
+import { SIGNAL_ABORTED } from '../../utils';
+
+/**
+ * Preserves (double) slashes earlier in the path, so this works better
+ * for URLs. From https://stackoverflow.com/a/46427607/4178400
+ * @param args parts of a path or URL to join.
+ */
+function joinUrlParts(...args: string[]) {
+  return args
+    .map((part, i) => {
+      if (i === 0) return part.trim().replace(/[/]*$/g, '');
+      return part.trim().replace(/(^[/]*|[/]*$)/g, '');
+    })
+    .filter(x => x.length)
+    .join('/');
+}
+
+class ReadOnlyStore {
+  async keys() {
+    return [];
+  }
+
+  async deleteItem() {
+    return false;
+  }
+
+  async setItem() {
+    console.warn('Cannot write to read-only store.');
+    return false;
+  }
+}
+
+export class FileStore
+  extends ReadOnlyStore
+  implements AsyncStore<ArrayBuffer> {
+  private _map: Map<string, File>;
+  private _rootPrefix: string;
+
+  constructor(fileMap: Map<string, File>, rootPrefix = '') {
+    super();
+    this._map = fileMap;
+    this._rootPrefix = rootPrefix;
+  }
+
+  private _key(key: string) {
+    return joinUrlParts(this._rootPrefix, key);
+  }
+
+  async getItem(key: string) {
+    const file = this._map.get(this._key(key));
+    if (!file) {
+      throw new KeyError(key);
+    }
+    const buffer = await file.arrayBuffer();
+    return buffer;
+  }
+
+  async containsItem(key: string) {
+    const path = this._key(key);
+    return this._map.has(path);
+  }
+}
+
+export class HTTPStore
+  extends ReadOnlyStore
+  implements AsyncStore<ArrayBuffer> {
+  private _signal: AbortSignal | undefined;
+
+  constructor(public url: string, public options?: RequestInit) {
+    super();
+    this._signal = undefined;
+  }
+
+  async getItem(key: string) {
+    const url = joinUrlParts(this.url, key);
+
+    /*
+     * If custom request options or a valid signal,
+     * pass the signal to the fetch request.
+     */
+    let options: undefined | RequestInit;
+    if (this.options || this._signal) {
+      options = { ...this.options, signal: this._signal };
+    }
+
+    const value = await fetch(url, options);
+
+    /*
+     * Throw our custom abort signal value to pick up in getTileData.
+     */
+    if (this._signal?.aborted) {
+      throw SIGNAL_ABORTED;
```

Comment:
You can throw any value. The goal is to just have something specific that bubbles up to the try/catch block in `getTileData`. In reality, this code rarely gets called because `fetch` throws a `DomException` that isn't caught here and deck.gl handles. If you add a console.log to the `catch` block of `getTileData` and log err you'll see that we actually don't need to throw anything special at all. It's just because Geotiff tries to handle the execption instead of letting it bubble up that we need something in the tiff source.

Reply:
I think Geotiff (and we by extension) may be moving to `DomException` as the default so I would not be opposed to making that the standard for the new loaders.

Reply:
To me, it is the "correct" to not have geotiff or any zarr code try to handle the error thrown by fetch when I signal is aborted. We wouldn't have to throw anything, and the DOMException thrown by the fetch request would bubble up to the try/catch is getTileData. It's actually kind of inconvenient that geotiff does something.

Reply:
The issue with geotiff was that one tile !== one request and multiple tiles could actually make the same request repeatedly and then abort them so the desired behavior wasn't clear to me.  I think what I'm saying is that soon Geotiff will however mock the error thrown after abort by throwing a `DomExcpetion` whose name is `AbortError` manually just as though it were one request being aborted in fetch (in which case it would be like what you are saying, as if there were no handling at all).  We are an upgrade away from that being the case so that was my motivation here (and in the tiff source as well), to move towards uniformity.

Reply:
Right, I understand one tile doesn't mean one request. But the first fetch request that is cancelled with throw a DOMException that will bubbled up to the caller if not caught. There is no case of "repeated requests" after one is cancelled (if you just let that exception bubble up).